apps/emqx_bpapi/src/emqx_bpapi_trans.erl

```diff
@@ -0,0 +1,200 @@
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+%% @hidden This parse transform generates BPAPI metadata function for
+%% a module, and helps dialyzer typechecking RPC calls
+-module(emqx_bpapi_trans).
+
+-export([parse_transform/2, format_error/1]).
+
+%%-define(debug, true).
+
+-define(META_FUN, bpapi_meta).
+
+-type semantics() :: call | cast.
+
+-record(s,
+        { api              :: atom()
+        , module           :: atom()
+        , version          :: non_neg_integer() | undefined
+        , introduced_in    :: emqx_bpapi:semver() | undefined
+        , deprecated_since :: emqx_bpapi:semver() | undefined
+        , targets = []     :: [{semantics(), emqx_bpapi:call(), emqx_bpapi:call()}]
+        , errors = []      :: [string()]
+        , file
+        }).
+
+format_error(invalid_name) ->
+  "BPAPI module name should follow <API>_proto_v<number> pattern";
+format_error(invalid_introduced_in) ->
+  "-introduced_in attribute should be present and its value should be a semver string";
+format_error(invalid_deprecated_since) ->
+  "value of -deprecated_since attribute should be a semver string";
+format_error({invalid_fun, Name, Arity}) ->
+  io_lib:format("malformed function ~p/~p. "
+                "BPAPI functions should have exactly one clause "
+                "and call (emqx_|e)rpc at the top level",
+                [Name, Arity]).
+
+parse_transform(Forms, _Options) ->
+  log("Original:~n~p", [Forms]),
+  State = #s{file = File} = lists:foldl(fun go/2, #s{}, Forms),
+  log("parse_trans state: ~p", [State]),
+  case check(State) of
+    [] ->
+      finalize(Forms, State);
+    Errors ->
+      {error, [{File, [{Line, ?MODULE, Msg} || {Line, Msg} <- Errors]}], []}
+  end.
+
+%% Scan erl_forms:
+go({attribute, _, file, {File, _}}, S) ->
+  S#s{file = File};
+go({attribute, Line, module, Mod}, S) ->
+  case emqx_bpapi:api_and_version(Mod) of
+    {ok, API, Vsn} -> S#s{api = API, version = Vsn, module = Mod};
+    error          -> push_err(Line, invalid_name, S)
+  end;
+go({attribute, _Line, introduced_in, Str}, S) ->
+  case is_list(Str) andalso emqx_bpapi:parse_semver(Str) of
+    {ok, Vsn} -> S#s{introduced_in = Vsn};
+    false     -> S %% Don't report error here, it's done in check/1
+  end;
+go({attribute, Line, deprecated_since, Str}, S) ->
+  case is_list(Str) andalso emqx_bpapi:parse_semver(Str) of
+    {ok, Vsn} -> S#s{deprecated_since = Vsn};
+    false     -> push_err(Line, invalid_deprecated_since, S)
+  end;
+go({function, Line, Name, Arity, Clauses}, S) ->
+  analyze_fun(Line, Name, Arity, Clauses, S);
+go(_, S) ->
+  S.
+
+check(#s{errors = Err0, introduced_in = II}) ->
+  [{none, invalid_introduced_in} || II =:= undefined] ++
+    Err0.
+
+finalize(Forms, S) ->
+  {Attrs, Funcs} = lists:splitwith(fun is_attribute/1, Forms),
```

Comment:
With `splitwith`, if attributes are placed in the middle of the file, they'll stay in `Funcs`.  Maybe `lists:partition` would cover that case?

Reply:
Attributes can't be placed in the middle of the file, since erlang won't compile such a file.